File: frontend/src/api/auditLog.ts
import type { AxiosResponse } from "axios";
import api from "./axios";

/**
 * Admin audit log entry. Mirrors the AuditLog JPA entity that maps to the
 * audit_logs table. Records who (actorId) did what (action) to which entity
 * (targetType/targetId), with optional JSON details and the originating IP.
 */
export interface AuditLogEntry {
    id: number;
    actorId: number | null;
    action: string;
    targetType: string | null;
    targetId: number | null;
    detailsJson: string | null;
    ipAddress: string | null;
    createdAt: string;
}

export interface SpringPage<T> {
    content: T[];
    totalElements: number;
    totalPages: number;
    number: number;
    size: number;
}

export interface AuditLogParams {
    action?: string;
    page?: number;
    size?: number;
}

export const getAuditLogs = (
    params: AuditLogParams,
): Promise<AxiosResponse<SpringPage<AuditLogEntry>>> =>
    api.get<SpringPage<AuditLogEntry>>("/admin/audit-log", { params });

File: backend/lernchih/src/main/java/com/richardjiang880/lernchih/model/AuditLog.java
package com.richardjiang880.lernchih.model;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

/**
 * JPA entity mapping to the {@code audit_logs} table (created in V12).
 *
 * Captures a single administrative action: who performed it ({@link #actorId}),
 * what was done ({@link #action}), what entity was targeted
 * ({@link #targetType} / {@link #targetId}), optional structured details
 * ({@link #detailsJson} as JSON text), and the originating IP address
 * ({@link #ipAddress}, added in V20260708220003).
 */
@Entity
@Table(name = "audit_logs")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class AuditLog {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "actor_id")
    private Long actorId;

    @Column(nullable = false)
    private String action;

    @Column(name = "target_type")
    private String targetType;

    @Column(name = "target_id")
    private Long targetId;

    @Column(name = "details_json", columnDefinition = "TEXT")
    private String detailsJson;

    @Column(name = "ip_address")
    private String ipAddress;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        if (createdAt == null) {
            createdAt = LocalDateTime.now();
        }
    }
}
